agent/retry_join.go

```diff
@@ -52,23 +52,32 @@ func newDiscover() (*discover.Discover, error) {
 	)
 }
 
-func retryJoinAddrs(disco *discover.Discover, cluster string, retryJoin []string, logger *log.Logger) []string {
+func retryJoinAddrs(disco *discover.Discover, cluster string, retryJoin []string, logger hclog.Logger) []string {
 	addrs := []string{}
 	if disco == nil {
 		return addrs
 	}
 	for _, addr := range retryJoin {
 		switch {
 		case strings.Contains(addr, "provider="):
-			servers, err := disco.Addrs(addr, logger)
+			servers, err := disco.Addrs(addr, logger.StandardLogger(&hclog.StandardLoggerOptions{
+				InferLevels: true,
+			}))
 			if err != nil {
 				if logger != nil {
-					logger.Printf("[ERR] agent: Cannot discover %s %s: %s", cluster, addr, err)
+					logger.Error("Cannot discover address",
+						"address", addr,
+						"cluster", cluster,
+						"error", err,
+					)
 				}
 			} else {
 				addrs = append(addrs, servers...)
 				if logger != nil {
-					logger.Printf("[INFO] agent: Discovered %s servers: %s", cluster, strings.Join(servers, " "))
+					logger.Info("Discovered servers",
+						"cluster", cluster,
```

Comment:
Isn't this `cluster` field already present due to the `With("cluster",...)` used when the derived logger is created?